lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,488 @@
+
+        while ($node !== null) {
+            if (! $node instanceof DOMElement) {
+                $node = $this->nextNode($node);
+                continue;
+            }
+
+            $heroImage = $this->detectImageWithDataHero($node);
+            if ($heroImage) {
+                $heroImages[] = $heroImage;
+            }
+
+            if (! $heroImageCandidate && count($heroImages) === 0) {
+                $heroImageCandidate = $this->detectHeroImageCandidate($node);
+            }
+            if (Amp::isTemplate($node)) {
+                // Ignore images inside templates.
+                $node = $this->skipNodeAndChildren($node);
+            } else {
+                $node = $this->nextNode($node);
+            }
+        }
+
+        // Optimize data-hero element if defined.
+        if (count($heroImages) > 0) {
+            return $heroImages;
+        }
+
+        // Fall back to auto-detected hero image if available.
+        if ($heroImageCandidate) {
+            return [$heroImageCandidate];
+        }
+
+        // No hero images to optimize.
+        return [];
+    }
+
+    /**
+     * Detect a hero image with the data-hero attribute.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image, or null if none detected.
+     */
+    private function detectImageWithDataHero(DOMElement $element)
+    {
+        if (
+            $element->tagName === Extension::IMAGE
+            && $element->hasAttribute(Attribute::DATA_HERO)
+        ) {
+            return new HeroImage(
+                $element->getAttribute(Attribute::SRC),
+                $element->getAttribute(Attribute::MEDIA),
+                $element->getAttribute(Attribute::SRCSET),
+                $element
+            );
+        }
+
+        if (
+            Amp::isAmpIframe($element)
+            && $element->hasAttribute(Attribute::DATA_HERO)
+        ) {
+            return $this->getPlaceholderImage($element);
+        }
+
+        return null;
+    }
+
+    /**
+     * Detect a hero image candidate.
+     *
+     * The hero image here can come from one of <amp-img>, <amp-video>, <amp-iframe>, <amp-video-iframe>.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidate(DOMElement $element)
+    {
+        if (
+            $element->hasAttribute(Attribute::LAYOUT)
+            && $element->getAttribute(Attribute::LAYOUT) === Layout::NODISPLAY
+        ) {
+            return null;
+        }
+
+        if ($element->tagName === Extension::IMAGE) {
+            return $this->detectHeroImageCandidateForAmpImg($element);
+        }
+
+        if ($element->tagName === EXTENSION::VIDEO) {
+            return $this->detectHeroImageCandidateForPosterImage($element);
+        }
+
+        if (Amp::isAmpIframe($element)) {
```

Comment:
This condition can be expanded to include _all_ embeds, including `amp-youtube`, `amp-vimeo`, etc.

Reply:
I'm unsure what the "all embeds" criteria actually is. As far as I understand, `placeholder` is a globally allowed attribute, but it can contain any HTML, not only images. Looking through the spec does not show anything I could use either.

In case you're referring to "oembeds", then I think we still need to be more selective, as an oembed can be something as simple as a link or a piece of text.

Right now, my thinking is that this has to be a hand-curated list of components that we assume could meet the following two conditions at the same time:
- be shown above the fold in the initial page render
- cause a large enough paint to be a likely candidate for LCS

Therefore, images, videos and oembeds for things we know take substantial space seem to make sense.

Reply:
This makes me wonder whether something like an `amp-ad` with a placeholder should be considered as well...
I guess if it fits the criteria, it is still worthwhile optimizing it...?

Reply:
Although `placeholder` is allowed everywhere, it doesn't seem to be properly used everywhere, though.
I just tested with an `amp-gallery` and didn't see the expected behavior.

Reply:
> I'm unsure what the "all embeds" criteria actually is.

Right, I mean any AMP component then. So any node whose `tagName` starts with `amp-`. In other words, only the custom elements that extend [`AMP.BaseElement`](https://github.com/ampproject/amphtml/blob/eb63b406b104f294bcdec71a39f64bbdbf7e611b/src/base-element.js#L108).

> Although `placeholder` is allowed everywhere, it doesn't seem to be properly used everywhere, though.

Yeah, I suppose would be up to the component to implement the appropriate `placeholder` behavior based on the component. If an author has added a `placeholder` to an AMP component, then we can assume that the author intends and verifies that the AMP component shows the placeholder while the component is loading.

> I just tested with an `amp-gallery` and didn't see the expected behavior.

Humm. What is an `amp-gallery`?

Reply:
Oh, sorry, I meant an `amp-carousel`.

Reply:
Looks like `amp-carousel` doesn't support `placeholder`. I suppose it would be the slides themselves that would have placeholders. So if you have an `amp-carousel` of `amp-youtube` elements, then you'd obtain the `amp-youtube > a[placeholder] > amp-img` for preloading. This makes sense to me as there's nothing in the carousel UI itself that would need to be preloaded.

Reply:
I tested the carousel with large images, and it then takes a lot of time to display, as it loads all of the referenced images first before starting to display the carousel. Durin the entire loading process, it shows a loading indicator.

I expected that adding a placeholder would allow you to display a smaller image or similar during the loading process, until all images are loaded. However, although you can see the placeholder being loaded on the Network tab, it is never displayed.

Reply:
That could be a defect/enhancement for `amp-carousel`.

Reply:
Yes, I thought so as well and will create an issue for that. As it is now, it is not really performant if you have big and/or many images.